lib/Dialect/FIRRTL/Transforms/ModuleInliner.cpp

```diff
@@ -0,0 +1,394 @@
+//===- ModuleInliner.cpp - FIRRTL module inlining ---------------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file implements FIRRTL module instance inlining.
+//
+//===----------------------------------------------------------------------===//
+
+#include "./PassDetails.h"
+#include "circt/Dialect/FIRRTL/FIRRTLOps.h"
+#include "circt/Dialect/FIRRTL/FIRRTLTypes.h"
+#include "circt/Dialect/FIRRTL/FIRRTLVisitors.h"
+#include "circt/Dialect/FIRRTL/Passes.h"
+#include "circt/Support/LLVM.h"
+#include "mlir/IR/BlockAndValueMapping.h"
+
+using namespace circt;
+using namespace firrtl;
+
+//===----------------------------------------------------------------------===//
+// Module Inlining Support
+//===----------------------------------------------------------------------===//
+
+/// If this operation or any child operation has a name, add the prefix to that
+/// operation's name.
+static void rename(const Twine &prefix, Operation *op) {
```

Comment:
The pervasive use of twine arguments here makes me concerned about dangling pointer bugs that happen due to twine temporaries.  Would you lose much by having these take StringRef prefixes instead?

Reply:
I had a bug with this when I was working on it, took me a while to find it even though I knew what the issue was going to be.  I have updated it to use StringRefs :)